### PR TITLE
feat(status): add conditions and related objects

### PR DESCRIPTION
**Pull request by djzager**
This is what it would look like if the status logic was vendored in.

### CODE REVIEW

File: pkg/controller/knicluster/knicluster_controller.go [hunk 1]
```diff
@@ -127,13 +128,12 @@ func (r *ReconcileKNICluster) ensureOperatorGroup(instance *kniv1alpha1.KNIClust
 	// already exists - don't requeue
 	reqLogger.Info("OperatorGroup already exists", "OperatorGroup.Namespace", found.Namespace, "OperatorGroup.Name", found.Name)
 
+	objectRef, err := reference.GetReference(r.scheme, found)
```
Comment (mhrivnak) [inline pkg/controller/knicluster/knicluster_controller.go:132]: Nice. That was convenient.

File: pkg/controller/knicluster/knicluster_controller.go [hunk 2]
```diff
@@ -251,6 +279,28 @@ func (r *ReconcileKNICluster) Reconcile(request reconcile.Request) (reconcile.Re
...
+		conditions.SetStatusCondition(&instance.Status.Conditions, conditions.Condition{
+			Type:   conditions.ConditionDegraded,
+			Status: corev1.ConditionFalse,
+		})
+		conditions.SetStatusCondition(&instance.Status.Conditions, conditions.Condition{
+			Type:   conditions.ConditionUpgradeable,
+			Status: corev1.ConditionUnknown,
+		})
+
+		_ = r.client.Status().Update(context.TODO(), instance)
```
Comment (mhrivnak) [inline pkg/controller/knicluster/knicluster_controller.go]: Any reason to not check the error here?

File: pkg/controller/knicluster/knicluster_controller.go [hunk 3]
```diff
@@ -276,11 +326,38 @@ func (r *ReconcileKNICluster) Reconcile(request reconcile.Request) (reconcile.Re
 	} {
 		err = f(instance, reqLogger)
 		if err != nil {
+			conditions.SetStatusCondition(&instance.Status.Conditions, conditions.Condition{
+				Type:    conditions.ConditionDegraded,
+				Status:  corev1.ConditionTrue,
+				Reason:  "ReconcileFailed",
+				Message: fmt.Sprintf("Failed reconcilitation %v", err),
```
Comment (mhrivnak) [inline pkg/controller/knicluster/knicluster_controller.go]: s/reconcilitation/reconciliation/

File: pkg/controller/knicluster/knicluster_controller.go [hunk 4]
```diff
@@ -276,11 +326,38 @@ func (r *ReconcileKNICluster) Reconcile(request reconcile.Request) (reconcile.Re
...
+			})
+
+			_ = r.client.Status().Update(context.TODO(), instance)
```
Comment (mhrivnak) [inline pkg/controller/knicluster/knicluster_controller.go]: Any reason to not check the error here?

File: pkg/controller/knicluster/knicluster_controller.go [hunk 5]
```diff
@@ -251,6 +279,28 @@ func (r *ReconcileKNICluster) Reconcile(request reconcile.Request) (reconcile.Re
...
+			Type:   conditions.ConditionProgressing,
+			Status: corev1.ConditionTrue,
+		})
+		conditions.SetStatusCondition(&instance.Status.Conditions, conditions.Condition{
+			Type:   conditions.ConditionDegraded,
+			Status: corev1.ConditionFalse,
+		})
+		conditions.SetStatusCondition(&instance.Status.Conditions, conditions.Condition{
+			Type:   conditions.ConditionUpgradeable,
+			Status: corev1.ConditionUnknown,
```
Comment (mhrivnak) [inline pkg/controller/knicluster/knicluster_controller.go:299]: Should this intentionally remain in "Unknown" state? I don't see anything else changing its value.

Comment (djzager) [reply pkg/controller/knicluster/knicluster_controller.go:299]: I think I'm going to remove it for now.

Comment (djzager) [reply pkg/controller/knicluster/knicluster_controller.go:299]: Decided to mark it as upgradeable when reconcile completes successfully

**Review by mhrivnak — APPROVED** [review]
@djzager anything else you wanted to do here, or is this ready to merge?

**Comment by djzager** [comment]
I think it's good. Once the github.com/openshift/custom-resource-status is ready, I'll submit a PR to update

**Pull request merged**